`ParquetReader.projectedGeneric` does not work when selecting more than one column from a same group

Hello, reading a parquet file with a group is failing when we try to select more than one column from the same group using `ParquetReader.projectedGeneric`.
In this example, we try to only read the `id`, `bar.a` and `bar.c` fields:

```scala
import com.github.mjakubowski84.parquet4s.*
import org.apache.parquet.hadoop.ParquetFileWriter
import java.nio.file.Path as NioPath

case class Foo(id: Int, bar: Bar)
case class Bar(a: String, b: String, c: String)
case class Baz(a: String, c: String)

val foos = List(
  Foo(1, Bar("a1", "b1", "c1")),
  Foo(2, Bar("a2", "b2", "c2"))
)

val path = "test.parquet"

val options = ParquetWriter.Options(
  writeMode = ParquetFileWriter.Mode.OVERWRITE
)
val writer = IncrementalParquetWriter[Foo](path, options)

def readBaz(path: NioPath): Iterable[Baz] =
  ParquetReader
    .projectedGeneric(
      Col("id").as[Int].alias("id"),
      Col("bar.a").as[String].alias("a"),
      Col("bar.c").as[String].alias("c")
    )
    .read(Path(path))
    .map(_.as[Baz](ValueCodecConfiguration.Default))

@main
def main(): Unit =
  writer.write(foos)
  writer.close()
  readBaz(NioPath.of(path)).foreach(println)
```
And it fails to read with this error:

```
Exception in thread "main" org.apache.parquet.io.ParquetDecodingException: Can not read value at 1 in block 0 in file file:/Users/y.alvarez/Prog/repos/test/test.parquet
	at org.apache.parquet.hadoop.InternalParquetRecordReader.nextKeyValue(InternalParquetRecordReader.java:264)
	at org.apache.parquet.hadoop.ParquetReader.read(ParquetReader.java:132)
	at org.apache.parquet.hadoop.ParquetReader.read(ParquetReader.java:136)
	at com.github.mjakubowski84.parquet4s.ParquetIterator.preRead(ParquetIterator.scala:29)
	at com.github.mjakubowski84.parquet4s.ParquetIterator.hasNext(ParquetIterator.scala:34)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:601)
	at scala.collection.Iterator$ConcatIterator.advance$1(Iterator.scala:1182)
	at scala.collection.Iterator$ConcatIterator.hasNext(Iterator.scala:1188)
	at scala.collection.Iterator$$anon$9.hasNext(Iterator.scala:583)
	at scala.collection.immutable.List.prependedAll(List.scala:152)
	at scala.collection.immutable.List$.from(List.scala:684)
	at scala.collection.immutable.List$.from(List.scala:681)
	at scala.collection.IterableFactory$Delegate.from(Factory.scala:288)
	at scala.collection.immutable.Iterable$.from(Iterable.scala:35)
	at scala.collection.immutable.Iterable$.from(Iterable.scala:32)
	at scala.collection.IterableFactory$Delegate.from(Factory.scala:288)
	at scala.collection.IterableOps.map(Iterable.scala:682)
	at scala.collection.IterableOps.map$(Iterable.scala:682)
	at com.github.mjakubowski84.parquet4s.etl.CompoundParquetIterable.map(CompoundParquetIterable.scala:6)
	at test.Main$package$.readBaz(Main.scala:31)
	at test.Main$package$.main(Main.scala:37)
	at test.main.main(Main.scala:33)
Caused by: java.lang.IllegalArgumentException: Invalid column projection: "bar.c".
	at com.github.mjakubowski84.parquet4s.RootRowParquetRecordConverter.end$$anonfun$1(ParquetReadSupport.scala:165)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.Vector.foreach(Vector.scala:2122)
	at com.github.mjakubowski84.parquet4s.RootRowParquetRecordConverter.end(ParquetReadSupport.scala:167)
	at org.apache.parquet.io.RecordReaderImplementation.read(RecordReaderImplementation.java:419)
	at org.apache.parquet.hadoop.InternalParquetRecordReader.nextKeyValue(InternalParquetRecordReader.java:234)
	... 21 more
```

--- a/build.sbt
+++ b/build.sbt
@@ -14,8 +14,8 @@ val akkaLib  = ActorLibCross("-akka", "-akka")
 val pekkoLib = ActorLibCross("-pekko", "-pekko")
 
 ThisBuild / organization := "com.github.mjakubowski84"
-ThisBuild / version := "2.15.0-SNAPSHOT"
-ThisBuild / isSnapshot := true
+ThisBuild / version := "2.14.1"
+ThisBuild / isSnapshot := false
 ThisBuild / scalaVersion := twoThirteen
 
 ThisBuild / javacOptions ++= Seq("-source", "1.8", "-target", "1.8")

--- a/build.sbt
+++ b/build.sbt
@@ -83,7 +83,10 @@ lazy val core = (projectMatrix in file("core"))
   .jvmPlatform(
     scalaVersions = Seq(twoTwelve, twoThirteen),
     settings = Def.settings(
-      libraryDependencies ++= sparkDeps :+ ("com.chuusai" %% "shapeless" % shapelessVersion)
+      libraryDependencies ++= sparkDeps :+ ("com.chuusai" %% "shapeless" % shapelessVersion),
+      excludeDependencies ++= Seq(
+        ExclusionRule("org.apache.logging.log4j", "log4j-slf4j2-impl")
+      )
     )
   )
   .jvmPlatform(

--- a/core/src/it/scala/com/github/mjakubowski84/parquet4s/ProjectionItSpec.scala
+++ b/core/src/it/scala/com/github/mjakubowski84/parquet4s/ProjectionItSpec.scala
@@ -147,10 +147,11 @@ class ProjectionItSpec extends AnyFlatSpec with Matchers with BeforeAndAfterAll 
 
     val expectedNestedRecord = RowParquetRecord(
       "b" -> ListParquetRecord(
-        RowParquetRecord("x" -> 1.value),
-        RowParquetRecord("x" -> 2.value),
-        RowParquetRecord("x" -> 3.value)
-      )
+        RowParquetRecord("x" -> 1.value, "y" -> "a".value),
+        RowParquetRecord("x" -> 2.value, "y" -> "b".value),
+        RowParquetRecord("x" -> 3.value, "y" -> "c".value)
+      ),
+      "c" -> true.value
     )
 
     try records.toList should be(
@@ -162,6 +163,21 @@ class ProjectionItSpec extends AnyFlatSpec with Matchers with BeforeAndAfterAll 
         )
       )
     )
+    finally records.close()
+  }
+
+  it should "allow to project complex fields for different subfields" in {
+    val records = ParquetReader
+      .projectedGeneric(
+        Col("nested.b").as[List[FullElem]],
+        Col("nested.c").as[Boolean]
+      )
+      .read(complexFilePath)
+      .as[FullNested]
+
+    val expectedRecord = FullNested(b = List(FullElem(1, "a"), FullElem(2, "b"), FullElem(3, "c")), c = true)
+
+    try records.toList should be(List(expectedRecord))
     finally records.close()
   }
 

--- a/core/src/main/scala/com/github/mjakubowski84/parquet4s/ParquetReadSupport.scala
+++ b/core/src/main/scala/com/github/mjakubowski84/parquet4s/ParquetReadSupport.scala
@@ -152,7 +152,8 @@ abstract private class RowParquetRecordConverter(schema: GroupType)
 private class RootRowParquetRecordConverter(schema: GroupType, columnProjections: Seq[ColumnProjection])
     extends RowParquetRecordConverter(schema) {
 
-  private lazy val emptyProjectionRow = RowParquetRecord.emptyWithSchema(columnProjections.map(cp => cp.alias.getOrElse(cp.columnPath.elements.last)))
+  private lazy val emptyProjectionRow =
+    RowParquetRecord.emptyWithSchema(columnProjections.map(cp => cp.alias.getOrElse(cp.columnPath.elements.last)))
 
   override def end(): Unit =
     if (columnProjections.nonEmpty) {

--- a/core/src/main/scala/com/github/mjakubowski84/parquet4s/ParquetReadSupport.scala
+++ b/core/src/main/scala/com/github/mjakubowski84/parquet4s/ParquetReadSupport.scala
@@ -152,17 +152,18 @@ abstract private class RowParquetRecordConverter(schema: GroupType)
 private class RootRowParquetRecordConverter(schema: GroupType, columnProjections: Seq[ColumnProjection])
     extends RowParquetRecordConverter(schema) {
 
+  private lazy val emptyProjectionRow = RowParquetRecord.emptyWithSchema(columnProjections.map(cp => cp.alias.getOrElse(cp.columnPath.elements.last)))
+
   override def end(): Unit =
-    columnProjections.foreach { case ColumnProjection(columnPath, ordinal, aliasOpt) =>
-      record.get(columnPath) match {
-        case Some(value) if columnPath.elements.length > 1 =>
-          record = record.updated(ordinal, aliasOpt.getOrElse(columnPath.elements.last), value)
-        case Some(_) =>
-          aliasOpt.foreach { alias =>
-            record = record.rename(ordinal, alias)
+    if (columnProjections.nonEmpty) {
+      record = columnProjections.foldLeft(emptyProjectionRow) {
+        case (newRecord, ColumnProjection(columnPath, _, aliasOpt)) =>
+          record.get(columnPath) match {
+            case Some(value) =>
+              newRecord.add(aliasOpt.getOrElse(columnPath.elements.last), value)
+            case None =>
+              throw new IllegalArgumentException(s"""Invalid column projection: "$columnPath".""")
           }
-        case None =>
-          throw new IllegalArgumentException(s"""Invalid column projection: "$columnPath".""")
       }
     }
 

--- a/core/src/main/scala/com/github/mjakubowski84/parquet4s/Schema.scala
+++ b/core/src/main/scala/com/github/mjakubowski84/parquet4s/Schema.scala
@@ -24,8 +24,8 @@ object Message {
     Types.buildMessage().addFields(fields*).named(name.getOrElse(DefaultName))
 
   /** Merges the fields before creating a schema. Merge is done by unifying types of columns that are defined in a
-    * projection more than once. The first mentioned column of primitive type is chosen in the case of duplicates.
-    * Union of of member fields is executed in the case of complex types.
+    * projection more than once. The first mentioned column of primitive type is chosen in the case of duplicates. Union
+    * of of member fields is executed in the case of complex types.
     * @param fields
     *   fields to be merged and then used for defining the schema
     * @return
@@ -41,7 +41,7 @@ object Message {
         register.get(fieldName) match {
           case Some(group: GroupType) if !tpe.isPrimitive() =>
             val newMemberFields = mergeFields(group.getFields().asScala.toSeq ++ tpe.asGroupType().getFields().asScala)
-            val mergedGroup = group.withNewFields(newMemberFields.asJava)
+            val mergedGroup     = group.withNewFields(newMemberFields.asJava)
             register.updated(fieldName, mergedGroup) -> (merged.filterNot(_ == group) :+ mergedGroup)
           case Some(firstSeen) =>
             register -> (merged :+ firstSeen)

--- a/core/src/main/scala/com/github/mjakubowski84/parquet4s/Schema.scala
+++ b/core/src/main/scala/com/github/mjakubowski84/parquet4s/Schema.scala
@@ -1,7 +1,7 @@
 package com.github.mjakubowski84.parquet4s
 
 import com.github.mjakubowski84.parquet4s.SchemaDef.Meta
-import org.apache.parquet.schema.{Types, *}
+import org.apache.parquet.schema.*
 import org.apache.parquet.schema.LogicalTypeAnnotation.{
   DateLogicalTypeAnnotation,
   DecimalLogicalTypeAnnotation,
@@ -14,6 +14,7 @@ import org.apache.parquet.schema.Type.Repetition
 
 import scala.annotation.nowarn
 import scala.reflect.ClassTag
+import scala.jdk.CollectionConverters.*
 
 object Message {
 
@@ -22,8 +23,9 @@ object Message {
   def apply(name: Option[String], fields: Type*): MessageType =
     Types.buildMessage().addFields(fields*).named(name.getOrElse(DefaultName))
 
-  /** Merges the fields before creating a schema. Merge is done by unifying types of columns that are define in a
-    * projection more than once. Type of the first mentioned column is chosen for each duplicate.
+  /** Merges the fields before creating a schema. Merge is done by unifying types of columns that are defined in a
+    * projection more than once. The first mentioned column of primitive type is chosen in the case of duplicates.
+    * Union of of member fields is executed in the case of complex types.
     * @param fields
     *   fields to be merged and then used for defining the schema
     * @return
@@ -37,6 +39,10 @@ object Message {
       .foldLeft(Map.empty[String, Type], Vector.empty[Type]) { case ((register, merged), tpe) =>
         val fieldName = tpe.getName
         register.get(fieldName) match {
+          case Some(group: GroupType) if !tpe.isPrimitive() =>
+            val newMemberFields = mergeFields(group.getFields().asScala.toSeq ++ tpe.asGroupType().getFields().asScala)
+            val mergedGroup = group.withNewFields(newMemberFields.asJava)
+            register.updated(fieldName, mergedGroup) -> (merged.filterNot(_ == group) :+ mergedGroup)
           case Some(firstSeen) =>
             register -> (merged :+ firstSeen)
           case None =>

--- a/project/DependecyVersions.scala
+++ b/project/DependecyVersions.scala
@@ -6,7 +6,7 @@ object DependecyVersions {
   val slf4jVersion                 = "2.0.9"
   val logbackVersion               = "1.3.11" // stick to 1.3.x for JDK-8 compatibility
   val akkaVersion                  = "2.6.21" // non-licensed version
-  val fs2Version                   = "3.9.2"
+  val fs2Version                   = "3.9.3"
   val catsEffectVersion            = "3.5.2"
   val scalaCollectionCompatVersion = "2.11.0"
   val scalatestVersion             = "3.2.17"

